### PR TITLE
[pr] Module 2 review

### PR DESCRIPTION
**Pull request by Okenlyzavals**

### CODE REVIEW

File: dao/src/main/java/com/epam/ems/dao/impl/SqlGiftCertificateDao.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        template = new JdbcTemplate(dataSource);
+        this.criteriaQueryBuilder = criteriaQueryBuilder;
+        this.updateQueryBuilder = updateQueryBuilder;
+        this.mapper = mapper;
+    }
+
+    @Override
+    public Optional<GiftCertificate> retrieveById(long id) {
+        Optional<GiftCertificate> certificate =
+                template.query(SQL_SELECT_ID, mapper, id).stream().findAny();
```
Comment (KseniyaYudzichavaTR) [inline dao/src/main/java/com/epam/ems/dao/impl/SqlGiftCertificateDao.java:79]: do not use stream().findAny() as only 1 result or null will return

File: dao/src/main/java/com/epam/ems/dao/impl/SqlGiftCertificateDao.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        List<GiftCertificate> certs = template.query(SQL_SELECT, mapper);
+        for(GiftCertificate certificate : certs){
+            certificate.setTags(retrieveTags(certificate.getId()));
+        }
+        return certs;
+    }
+
+    @Override
+    @Transactional
+    public Long create(GiftCertificate entity) {
```
Comment (KseniyaYudzichavaTR) [inline dao/src/main/java/com/epam/ems/dao/impl/SqlGiftCertificateDao.java:95]: Just explain if tag is new and certificate is new how tag will be created

File: dao/src/main/java/com/epam/ems/dao/impl/SqlGiftCertificateDao.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        MultiValueMap<String, List<Object>> queries = updateQueryBuilder.parse(entity, oldCertificate);
+        for(String key : queries.keySet()){
+            for(List<Object> params : queries.get(key)){
+                template.update(key, params.toArray());
+            }
+        }
+    }
+
+    @Override
+    public List<Tag> retrieveTags(long id){
```
Comment (KseniyaYudzichavaTR) [inline dao/src/main/java/com/epam/ems/dao/impl/SqlGiftCertificateDao.java:154]: why a method with tags is inside of Certificate class?

File: dao/src/main/java/com/epam/ems/dao/impl/SqlGiftCertificateDao.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+            statement.setObject(1,entity.getName());
+            statement.setObject(2,entity.getDescription());
+            statement.setObject(3,entity.getPrice());
+            statement.setObject(4,entity.getDuration());
+            statement.setObject(5, Timestamp.valueOf(entity.getCreateDate()));
+            statement.setObject(6, Timestamp.valueOf(entity.getLastUpdateDate()));
+            return statement;
+        },holder);
+        Long key = holder.getKey().longValue();
+        for(Tag tag : entity.getTags()){
```
Comment (KseniyaYudzichavaTR) [inline dao/src/main/java/com/epam/ems/dao/impl/SqlGiftCertificateDao.java:110]: I do not like this loop inside - it should be in a separate method

File: dao/src/main/java/com/epam/ems/dao/impl/SqlGiftCertificateDao.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+        return certs;
+    }
+
+    @Override
+    @Transactional
+    public void update(GiftCertificate entity) {
+
+        GiftCertificate oldCertificate = retrieveById(entity.getId()).get();
+        MultiValueMap<String, List<Object>> queries = updateQueryBuilder.parse(entity, oldCertificate);
+        for(String key : queries.keySet()){
```
Comment (KseniyaYudzichavaTR) [inline dao/src/main/java/com/epam/ems/dao/impl/SqlGiftCertificateDao.java:146]: double loop in one method - is it ok for clean code?

File: dao/src/main/java/com/epam/ems/dao/impl/SqlGiftCertificateDao.java [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+    @Override
+    public List<GiftCertificate> retrieveByCriteria(Criteria criteria) {
+        criteriaQueryBuilder.parse(criteria);
+        List<GiftCertificate> certs
+                = template.query(
+                        criteriaQueryBuilder.getQuery(),
+                        mapper,
+                        criteriaQueryBuilder.getParams());
+        for(GiftCertificate certificate : certs){
```
Comment (KseniyaYudzichavaTR) [inline dao/src/main/java/com/epam/ems/dao/impl/SqlGiftCertificateDao.java:134]: duplication of code - send it to a private method

File: dao/src/main/java/com/epam/ems/dao/impl/SqlTagDao.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        template = new JdbcTemplate(dataSource);
+        this.tagRowMapper = rowMapper;
+    }
+
+    @Override
+    public Optional<Tag> retrieveById(long id) {
+        return template.query(
+                SQL_SELECT_ID,
+                tagRowMapper,
+                id).stream().findAny();
```
Comment (KseniyaYudzichavaTR) [inline dao/src/main/java/com/epam/ems/dao/impl/SqlTagDao.java:55]: stream().findAny()

File: dao/src/main/java/com/epam/ems/dao/impl/SqlTagDao.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    @Override
+    public void delete(Tag entity) {
+        template.update(SQL_DELETE, entity.getId());
+    }
+
+    @Override
+    public Optional<Tag> findByName(String name) {
+        return template.query(SQL_SELECT_NAME, tagRowMapper, name).stream().findAny();
```
Comment (KseniyaYudzichavaTR) [inline dao/src/main/java/com/epam/ems/dao/impl/SqlTagDao.java:86]: name is unique and..... stream().findAny()

File: dao/src/main/java/com/epam/ems/dao/querybuilder/UpdateQueryBuilder.java [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+        }
+    }
+
+    private MultiValueMap<String, List<Object>> getTagUpdateQueries(){
+        MultiValueMap<String, List<Object>> result = new LinkedMultiValueMap<>();
+
+        if(newCertificate.getTags() == null){
+            return result;
+        }
+        HashSet<Tag> oldTags = new HashSet<>(oldCertificate.getTags());
```
Comment (KseniyaYudzichavaTR) [inline dao/src/main/java/com/epam/ems/dao/querybuilder/UpdateQueryBuilder.java:83]: HashSet<Tag> oldTags = new HashSet<>()?????? why such implementation is used. Explain please

File: dao/src/main/java/com/epam/ems/dao/querybuilder/UpdateQueryBuilder.java [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+        HashSet<Tag> newTags = new HashSet<>(newCertificate.getTags());
+
+        HashSet<Tag> temp = new HashSet<>(oldTags);
+
+        temp.retainAll(newTags);
+        oldTags.removeAll(temp);
+        newTags.removeAll(temp);
+        for(Tag tag : oldTags){
```
Comment (KseniyaYudzichavaTR) [inline dao/src/main/java/com/epam/ems/dao/querybuilder/UpdateQueryBuilder.java:91]: ??? Strange logic and too complicated for 1 method

File: dao/src/main/java/com/epam/ems/dao/impl/SqlGiftCertificateDao.java (hunk 2 above)
Comment (Okenlyzavals) [reply dao/src/main/java/com/epam/ems/dao/impl/SqlGiftCertificateDao.java:95]: I've been creating new tags on service layer, but yes indeed, this process should be located in DAL, within single transaction; creation of new tags during certificate update will be transfered to DAL as well.

File: dao/src/main/java/com/epam/ems/dao/querybuilder/UpdateQueryBuilder.java (hunk 9 above)
Comment (Okenlyzavals) [reply dao/src/main/java/com/epam/ems/dao/querybuilder/UpdateQueryBuilder.java:83]: My mistake, changed it to Set<Tag> = new HashSet<>. The logic behind this action, if that concerns you, lies in finding tags that are common between an instance of certificate that is within the DB and an instance that is passed from service to be updated, so there will be less queries for DB.

**Pull request closed without merging**